vulcanai2/models/basenetwork.py

```diff
@@ -140,6 +141,8 @@ def __init__(self, name, config, in_dim=None, save_path=None,
         # Compute self.out_dim of the network
         self.out_dim = self._get_out_dim()
 
+        self.device = torch.device(device if torch.cuda.is_available() else "cpu")
```

Comment:
if cuda is available but the user forgets to specify the device it will still set it to the `cpu`? perhaps we should default `device='cuda:0'` instead @priyatharsan ?

Reply:
if cuda is not available then it will default to cpu anyway

Reply:
hmm.. but what if its the case, if the user still would like to stick with `cpu`

Reply:
they can always override `model.device = 'cpu'` after they've created the network. they would explicitly need to specify using the cpu if a cpu is available

Reply:
Why don't we have the setter for self.device written so that anytime the user changes self.device it calls all the necessary self.to(self.device) etc statements? Or is there something I'm missing?

Reply:
I think that can work. @priyatharsan ?

Reply:
also should it extend to all `self.input_networks` as well? similar to how when you call `freeze` you have the option to also freeze incoming networks? @priyatharsan

Reply:
I will give it a shot on setter today. That shouldn't be a problem. 

And we don't have to extend to all `self.input_networks`. The network is a nn.ModuleList (currently in this branch) so it supports .to(device) so setting it will, by default, switches the device internally to its own input_networks. I need to check how it reacts with nn.ModuleDict. However, those input_networks will be, individually, in the default device (cpu). Eg: setting `multi_input_conv_3D` to gpu will set itself and its input_networks to gpu but not set `conv_3D` to gpu unless we specify them manually. Do you get me?

Reply:
im pretty sure ModuleList and ModuleDict work the same way in terms of the `.to` functionality. Just to clarify your example, if `conv_3D` is an input to the `multi_input_conv_3D`, and you set the device of the latter to gpu, then it will also change the former?

Reply:
no it won't explicitly but internally yes. To be more clear, if we set `self.to(device`) it sets `self.network` and `self.input_networks = nn.ModuleList(list of input_networks that are nn.Modules)` `.to(device)` but not individually on each individual `input_networks`. Does that makes sense?

Reply:
we will discuss this after the meeting today. If its confusing 😅

Reply:
I see. But I'm assuming that in addition to just the ModuleList the embedded networks will also be run on the specified device? maybe we should explicitly set it instead of the implicit assumption?

Reply:
>I'm assuming that in addition to just the ModuleList the embedded networks will also be run on the specified device?
yep! they run on the specified device assigned to nn.ModuleList.

>maybe we should explicitly set it instead of the implicit assumption?
hmm... we can do that by looping over individual input_networks just like we did before when used `list` object instead of `nn.ModuleList`, remember `_to_cuda` func?. That's an additional step, but is that necessary?

Reply:
we can confirm that simply setting the `ModuleList` to a device that the internal networks also get assigned to that device during run time. that would clear up my concerns

Reply:
Yep it works. The current version in this branch trains in my gpu without needing to individually switch the device for the input_networks. Is it possible you can set up the ml server to cuda ready today?